app/src/main/java/io/apicurio/registry/rules/compatibility/CompatibilityRuleExecutor.java

```diff
@@ -56,14 +57,16 @@ public void execute(RuleContext context) throws RuleViolationException {
         CompatibilityChecker checker = provider.getCompatibilityChecker();
         List<ContentHandle> existingArtifacts = context.getCurrentContent() != null
             ? singletonList(context.getCurrentContent()) : emptyList();
-        if (!checker.isCompatibleWith(
-            level,
-            existingArtifacts,
-            context.getUpdatedContent())
-        ) {
-            throw new RuleViolationException(String.format("Incompatible artifact: %s [%s]",
-                context.getArtifactId(), context.getArtifactType()),
-                RuleType.COMPATIBILITY, context.getConfiguration());
+
+        CompatibilityExecutionResult compatibilityExecutionResult = checker.getIncompatibleDifferences(
+                level,
+                existingArtifacts,
+                context.getUpdatedContent());
+        if (!compatibilityExecutionResult.isCompatible()) {
+            throw new RuleViolationException(String.format("Incompatible artifact: %s [%s] {%s}",
+                context.getArtifactId(), context.getArtifactType(),
+                    Arrays.toString(compatibilityExecutionResult.getIncompatibleDifferences().toArray())),
```

Comment:
I don't think we should include the set of differences in the exception message.  Those details will be included in the exception and then communicated back via the changes to the `Error` object.  I think having them in the message as well will only clutter the message.  Perhaps include the **number** of problems found in the message instead.

Reply:
sounds good